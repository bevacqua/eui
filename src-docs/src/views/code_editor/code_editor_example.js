import React from 'react';

import { renderToHtml } from '../../services';

import {
  GuidePage,
  GuideSection,
  GuideSectionTypes,
} from '../../components';

import {
  EuiCode,
} from '../../../../src/components';

import CodeEditor from './code_editor';
const codeEditorSource = require('!!raw-loader!./code_editor');
const codeEditorHtml = renderToHtml(CodeEditor);

export default props => (
  <GuidePage title={props.route.name}>
    <GuideSection
      title="Code Editor"
      source={[{
        type: GuideSectionTypes.JS,
        code: codeEditorSource,
      }, {
        type: GuideSectionTypes.HTML,
        code: codeEditorHtml,
      }]}
      text={
        <div>
          <p>
            The <EuiCode>EuiCodeEditor</EuiCode> component is a wrapper around <EuiCode>codemirror</EuiCode>.
          </p>
          <p>
            All parameters, that you specify are passed down to the
            underlying <EuiCode>CodeMirror</EuiCode> component.
          </p>
        </div>
      }
      demo={
        <CodeEditor />
      }
    />
  </GuidePage>
);

