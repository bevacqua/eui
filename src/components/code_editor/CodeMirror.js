// Fork of https://github.com/JedWatson/react-codemirror (MIT license)

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import codeMirror from 'codemirror';
import { differenceWith, isEqual, noop, toPairs } from 'lodash';

class CodeMirror extends Component {
  componentDidMount() {
    const { options, value } = this.props;

    this.codeMirror = codeMirror.fromTextArea(this._textareaNode, options);
    this.codeMirror.on(`change`, this.codemirrorValueChanged.bind(this));
    this.codeMirror.on(`focus`, this.focusChanged.bind(this, true));
    this.codeMirror.on(`blur`, this.focusChanged.bind(this, false));
    this.codeMirror.on(`scroll`, this.scrollChanged.bind(this));
    this.codeMirror.setValue(value);
  }

  componentWillReceiveProps(nextProps) {
    this.updateValue(nextProps.value);
    this.updateName(nextProps.name);
    this.updateOptions(nextProps.options);
    this.updateClassName(nextProps.className);
  }

  shouldComponentUpdate() {
    return !(this._textareaNode && this._containerNode);
  }

  componentWillUnmount() {
    if (this.codeMirror) {
      this.codeMirror.toTextArea();
    }
  }

  render() {
    const { path, value } = this.props;
    return (
      <div
        ref={c => { this._containerNode = c; }}
        className={this.computeClassName()}
      >
        <textarea
          ref={c => { this._textareaNode = c; }}
          name={path}
          defaultValue={value}
          autoComplete="off"
        />
      </div>
    );
  }

  focus() {
    if (this.codeMirror) {
      this.codeMirror.focus();
    }
  }

  focusChanged(focused) {
    this.updateClassName();
    this.props.onFocusChange(focused);
  }

  scrollChanged(cm) {
    this.props.onScroll(cm.getScrollInfo());
  }

  codemirrorValueChanged(doc, change) {
    if (change.origin !== `setValue`) {
      this.props.onChange(doc.getValue());
    }
  }

  computeClassName(userClassName) {
    const cm = this.codeMirror;

    return cx(
      `euiCodeMirror`,
      userClassName || this.props.className,
      { 'euiCodeMirror--focused': cm && cm.hasFocus() }
    );
  }

  updateValue(next = ``) {
    const cm = this.codeMirror;
    if (!cm) return;
    if (next === cm.getValue()) return;

    cm.setValue(next);
  }

  updateName(next) {
    const textarea = this._textareaNode;
    if (!textarea) return;
    if (next === textarea.getAttribute(`name`)) return;

    if (next) {
      textarea.setAttribute(`name`, next);
    } else {
      textarea.removeAttribute(`name`);
    }
  }

  updateOptions(next) {
    const currentPairs = toPairs(this.props.options);
    const nextPairs = toPairs(next);
    const pairIsEqual = (a, b) => isEqual(a[1], b[1]);
    const updatedPairs = differenceWith(nextPairs, currentPairs, pairIsEqual);

    updatedPairs.forEach(pair =>
      this.codeMirror.setOption(...pair)
    );
  }

  updateClassName(userClassName) {
    const cm = this.codeMirror;
    const container = this._containerNode;
    if (!cm || !container) return;

    const className = this.computeClassName(userClassName);

    if (container.className !== className) {
      container.className = className;
    }
  }
}

CodeMirror.defaultProps = {
  onFocusChange: noop,
  onScroll: noop,
  onChange: noop,
  value: ``
};

CodeMirror.propTypes = {
  onChange: PropTypes.func,
  onFocusChange: PropTypes.func,
  onScroll: PropTypes.func,
  options: PropTypes.object,
  path: PropTypes.string,
  value: PropTypes.string,
  className: PropTypes.any
};

export default CodeMirror;
