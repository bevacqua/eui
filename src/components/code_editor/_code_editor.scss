@import '~codemirror/lib/codemirror.css';
@import '~codemirror/theme/neat.css';
@import '~codemirror/theme/dracula.css';

.euiCodeMirror {
  max-width: 900px;
}
