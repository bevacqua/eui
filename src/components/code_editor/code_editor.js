import React, { Component } from 'react';
import PropTypes from 'prop-types';

import CodeMirror from './CodeMirror';

import 'codemirror/mode/javascript/javascript';
import 'codemirror/mode/yaml/yaml';

const languages = {
  json: `application/json`,
  yaml: `text/x-yaml`,
  text: `text/plain`
};

const themes = {
  light: `neat`,
  dark: `dracula`
};

const COLOR_NAMES = Object.keys(themes);

export class EuiCodeEditor extends Component {
  codeMirrorRef = (codeMirror) => {
    if (codeMirror) {
      this.codeMirror = codeMirror;
    }
  };

  render() {
    const {
      color,
      language,
      onBlur, // eslint-disable-line no-unused-vars
      ...rest
    } = this.props;

    const codeMirrorOptions = {
      mode: languages[language] || language,
      theme: themes[color],
      indentUnit: 2,
      tabSize: 2,
      smartIndent: true,
      lineNumbers: false,
      electricChars: true
    };

    return (
      <div className="euiCodeEditorWrapper">
        <CodeMirror
          ref={this.codeMirrorRef}
          options={codeMirrorOptions}
          {... rest}
        />
      </div>
    );
  }
}

EuiCodeEditor.propTypes = {
  color: PropTypes.oneOf(COLOR_NAMES),
  width: PropTypes.string,
  height: PropTypes.string,
  onBlur: PropTypes.func,
};
