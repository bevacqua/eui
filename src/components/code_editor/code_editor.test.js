import React from 'react';
import { mount } from 'enzyme';
import { EuiCodeEditor } from './code_editor';
import {
  requiredProps,
  takeMountedSnapshot,
} from '../../test';

jest.mock('./CodeMirror', () => `CodeMirror`);

describe('EuiCodeEditor', () => {
  test('is rendered', () => {
    const component = mount(<EuiCodeEditor {...requiredProps}/>);
    expect(takeMountedSnapshot(component)).toMatchSnapshot();
  });
});
