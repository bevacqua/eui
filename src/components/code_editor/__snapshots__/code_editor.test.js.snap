// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiCodeEditor is rendered 1`] = `
<div
  class="euiCodeEditorWrapper"
>
  <codemirror
    aria-label="aria-label"
    class="testClass1 testClass2"
    data-test-subj="test subject string"
    options="[object Object]"
  />
</div>
`;
